pkg/kpipelines/syntax/pipeline.go

```diff
@@ -0,0 +1,987 @@
+
+	Workspace *string `yaml:"workspace,omitempty"`
+}
+
+type Step struct {
+	// One of command or step is required.
+	Command string `yaml:"command,omitempty"`
+	// args is optional, but only allowed with command
+	Arguments []string `yaml:"args,omitempty"`
+	// dir is optional, but only allowed with command. Refers to subdirectory of workspace
+	Dir string `yaml:"dir,omitempty"`
+
+	Step string `yaml:"step,omitempty"`
+	// options is optional, but only allowed with step
+	// Also, we'll need to do some magic to do type verification during translation - i.e., this step wants a number
+	// for this option, so translate the string value for that option to a number.
+	Options map[string]string `yaml:"options,omitempty"`
+
+	// agent can be overridden on a step
+	Agent Agent `yaml:"agent,omitempty"`
+}
+
+type Stage struct {
+	Name        string       `yaml:"name"`
+	Agent       Agent        `yaml:"agent,omitempty"`
+	Options     StageOptions `yaml:"options,omitempty"`
+	Environment []EnvVar     `yaml:"environment,omitempty"`
+	Steps       []Step       `yaml:"steps,omitempty"`
+	Stages      []Stage      `yaml:"stages,omitempty"`
+	Parallel    []Stage      `yaml:"parallel,omitempty"`
+	Post        []Post       `yaml:"post,omitempty"`
+}
+
+type PostCondition string
+
+// Probably extensible down the road
+const (
+	PostConditionSuccess PostCondition = "success"
+	PostConditionFailure PostCondition = "failure"
+	PostConditionAlways  PostCondition = "always"
+)
+
+var AllPostConditions = []PostCondition{PostConditionAlways, PostConditionSuccess, PostConditionFailure}
+
+// TODO: Conditional execution of something after a Task or Pipeline completes is not yet implemented
+type Post struct {
+	Condition PostCondition `yaml:"condition"`
+	Actions   []PostAction  `yaml:"actions"`
+}
+
+// TODO: Notifications are not yet supported in Build Pipeline per se.
+type PostAction struct {
+	Name string `yaml:"name"`
+	// Also, we'll need to do some magic to do type verification during translation - i.e., this action wants a number
+	// for this option, so translate the string value for that option to a number.
+	Options map[string]string `yaml:"options,omitempty"`
+}
+
+var _ apis.Validatable = (*Jenkinsfile)(nil)
+var _ apis.Defaultable = (*Jenkinsfile)(nil)
+
+func (s *Stage) TaskName() string {
+	return strings.ToLower(strings.NewReplacer(" ", "-").Replace(s.Name))
+}
+
+// Task/Step names need to be RFC 1035/1123 compliant DNS labels, so we mangle
+// them to make them compliant. Results should match the following regex and be
+// no more than 63 characters long:
+//     [a-z]([-a-z0-9]*[a-z0-9])?
+// cf. https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+// body is assumed to have at least one ASCII letter.
+// suffix is assumed to be alphanumeric and non-empty.
+// TODO: Combine with kube.ToValidName (that function needs to handle lengths)
+func MangleToRfc1035Label(body string, suffix string) string {
```

Comment:
Happy to try to combine this with `kube.toValidName` by adding some length enforcement to that function if desired.